File: rd-net/RdFramework/Impl/LightSingleContextHandler.cs
using System;
using JetBrains.Rd.Base;
using JetBrains.Serialization;

#nullable disable

namespace JetBrains.Rd.Impl
{
  internal class LightSingleContextHandler<T> : ISingleContextHandler<T>
  {
    public RdContext<T> Context { get; }

    public RdContextBase ContextBase => Context;

    public LightSingleContextHandler(RdContext<T> context)
    {
      Context = context;
    }


    public object ReadValueBoxed(SerializationCtx context, UnsafeReader reader)
    {
      return ReadValue(context, reader);
    }

    public void WriteValue(SerializationCtx context, UnsafeWriter writer)
    {
      var value = Context.Value;
      writer.Write(value != null);
      if (value != null)
        Context.WriteDelegate(context, writer, value);
    }

    public void RegisterValueInValueSet()
    {
      // no-op
    }

    public T ReadValue(SerializationCtx context, UnsafeReader reader)
    {
      var hasValue = reader.ReadBool();
      if (!hasValue)
        return default;
      return Context.ReadDelegate(context, reader);
    }
  }
}

File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/impl/LightSingleContextHandler.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.IRdBindable
import com.jetbrains.rd.framework.base.ISingleContextHandler
import com.jetbrains.rd.framework.base.RdBindableBase

internal class LightSingleContextHandler<T: Any>(override val context: RdContext<T>) : RdBindableBase(), ISingleContextHandler<T> {
    override fun deepClone(): IRdBindable {
        error("This may not be cloned")
    }

    @Suppress("UNCHECKED_CAST")
    override fun writeValue(ctx: SerializationCtx, buffer: AbstractBuffer) {
        val value = context.value
        if(value == null)
            buffer.writeBoolean(false)
        else {
            buffer.writeBoolean(true)
            context.serializer.write(ctx, buffer, value)
        }
    }

    override fun readValue(ctx: SerializationCtx, buffer: AbstractBuffer): T? {
        val hasValue = buffer.readBoolean()
        if(!hasValue) return null
        return context.serializer.read(ctx, buffer)
    }

    override fun registerValueInValueSet() = Unit
}